clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -207,12 +212,14 @@ void run(long now) {
                     this.accumulator.mutePartition(batch.topicPartition);
             }
         }
-
-        List<RecordBatch> expiredBatches = this.accumulator.abortExpiredBatches(this.requestTimeout, now);
-        // update sensors
-        for (RecordBatch expiredBatch : expiredBatches)
-            this.sensors.recordErrors(expiredBatch.topicPartition.topic(), expiredBatch.recordCount);
-
+        boolean stale = isMetadataStale(now, metadata, this.metadataStaleMs);
+        if (stale || !result.unknownLeaderTopics.isEmpty()) {
```

Comment:
The condition in Sender.java is a quick check to avoid expensive iteration in `abortExpiredBatches` when we can. I.e., when metadata is fresh and leaders for all partitions are known there's no reason to drop into the iteration in `abortExpiredBatches`. We'll not expire any batches in that case.